### PR TITLE
handle count all correctly when distinct is passed in as a selected value

### PR DESCRIPTION
**Pull request by twitnithegirl**
### Summary

This addresses this issue https://github.com/rails/rails/issues/33345

TL;DR of the issue is that when you run something like `User.select("DISTINCT users.id").join(:event).count` the query forms correctly and you get the correct value. However if you do the same call but pass in `:all" to `#count()` the request is built with the distinct call missing.

### CODE REVIEW

**Comment by twitnithegirl** [comment]
Can someone with power restart [this job](https://travis-ci.org/rails/rails/jobs/404589905)? Failed on an apt-get install but all the other builds passed so I think it's a fluke failure.

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+*   Don't ignore `DISTICT` selector value when `:all` is passed into `#count`.
```
Comment (JulienItard) [inline activerecord/CHANGELOG.md]: You have a typo on `DISTINCT`

(It's my first "contribution" on Rails project \o/ !)

Comment (twitnithegirl) [reply activerecord/CHANGELOG.md]: ugh thanks. I typoed it in the code too originally. 🤦‍♀️

**Comment by azbshiri** [comment]
@twitnithegirl would you squash your commits into one?

**Comment by twitnithegirl** [comment]
@azbshiri i can but TBH if the only reason you want me to is for the post-merge log history to be neat, it is probably less complicated (and less likely to result in a mess) for whoever merges this in to just hit "Squash and Merge" than for me to rebase with 40 commits between my first and last change in log.

**Comment by twitnithegirl** [comment]
@azbshiri sure, here ya go.

TBH, though, I think it's probably cleaner to just do `Squash and Merge` as whoever is merging these in. Less likely to result in weird history.

File: activerecord/lib/active_record/relation/calculations.rb [hunk 2]
```diff
@@ -237,7 +237,11 @@ def perform_calculation(operation, column_name)
 
         # If #count is used with #distinct (i.e. `relation.distinct.count`) it is
         # considered distinct.
-        distinct = distinct_value
+        if select_values.any? && select_values.first =~ /\s*DISTINCT[\s(]+/i
```
Comment (kaspth) [inline activerecord/lib/active_record/relation/calculations.rb]: Maybe we could simplify this to `distinct = select_values.first&.match?(/\s*DISTINCT[\s(]+/i) || distinct_value`?

As we prefer `match?` to `=~`. Also is it `select_values.first` and not `select_values.any? { |v| v.match?(…)`?

File: activerecord/test/cases/calculations_test.rb [hunk 3]
```diff
@@ -248,6 +248,10 @@ def test_count_with_eager_loading_and_custom_order_and_distinct
     assert_queries(1) { assert_equal 11, posts.count(:all) }
   end
 
+  def test_count_all_with_distinct_selected_value_and_join
+    assert_equal 5, Account.select("DISTINCT accounts.id").joins(:firm).count(:all)
```
Comment (kaspth) [inline activerecord/test/cases/calculations_test.rb]: Just out of curiosity: would `Account.distinct.joins(:firm).count` also return 5?

Comment (twitnithegirl) [reply activerecord/test/cases/calculations_test.rb]: Thanks for this great feedback. I'll double check this and address your other comment above.

**Comment by twitnithegirl** [comment]
@kaspth I wrote a new test. the old test wasn't definitive. with the new test if you don't do `"DISTINCT"` or `.distinct` you get 6 but with you get 4. I ran the test with and without my change and it went red without and green with.

I also used `match?` instead of `=~`. I went with `.try()` instead of `&.` because the tests on 504 and 170 didn't work with the `&.` syntax but did work this way. I really don't know why  ¯\_(ツ)_/¯

**Comment by matthewd** [comment]
While anything that looks like SQL-parsing makes me a bit nervous, my main doubt here is that I'm concerned this breaks `count(:all)`'s contract, which is to override `select_values`.

It would clearly give the answer expected in the original issue, but it also (I believe?) changes other cases into exceptions: `User.select("DISTINCT ON (id) id, name").count(:all)` doesn't currently give a "distinctive" count, but it also doesn't raise.

In general, I'm not sure it's safe for us to second-guess the caller's intent when `count(:all)` is already supposed to be the escape-hatch: if they meant `COUNT(#{select_values})`, that's `count`'s default behaviour. :confused:

**Comment by twitnithegirl** [comment]
Thanks for the comments @matthewd Do you think the issue (and PR) should just be closed or is there a different route we could pursue?

**Comment by kamipo** [comment]
Closing in favor of #35361.

**Pull request closed without merging**